WordPress/gutenberg
# ColorPicker: improve UX of dragging the handle when in popover on top of the editor

Fixes #49267
Alternative approach to #54164

## What?


Fixes an issue with dragging the handle of the color picker when the `ColorPicker` component is rendered inside a `Popover` on top of the editor canvas (an `iframe`)

## Why?


The current UX is poor and leads to a frustrating interaction when the color picker closes unexpectedly in the middle of dragging its handle, should the user move the cursor on top of the editor canvas.

The reason why this bug is happening in the first place is because of how pointer events behave when moving the cursor over an `iframe`. By default, `iframe`s would absorb all pointer events — and therefore the color picker handle wouldn't update at all when moving the pointer on top of the editor canvas.

But the editor canvas' iframe has been tweaked to bubble certain events (including mouse move), and therefore uses don't experience the behaviour explained in the paragraph above. Instead, it looks like the cursor handle isn't updated as smoothly, and (even worse), moving the cursor while dragging on top of the editor canvas can result in the `iframe` itself receiving focus, which in turn causes the `Popover` around the `ColorPicker` to automatically close (it closes when losing focus).

## How?


The [approach](https://github.com/WordPress/gutenberg/pull/54164#issuecomment-1747700127) tried in #54164 works, but it results in many API changes to components, an extra context being added and in general quite a lot of code.

This solution, instead, tries to prevent the problem from happening by working solely at `@wordpress/components` level.

Basically, whenever a `ColorPicker` is rendered inside a `Popover` component, the `Popover` component will render an additional backdrop element _only while the color picker handle is being dragged_. The backdrop serves as a pointer events "trap", avoiding them from ever reaching the underlying `iframe` and therefore causing the bug described above.

A better solution, in the future, could be to allow `Popover` to expose a `modal` prop (similarly to how the `Popover` component from `ariakit` [already does](https://ariakit.org/reference/popover#modal)), which would basically implement this same solution but in a more robust way.

## Testing Instructions





1. Open the site editor
2. Select a paragraph block
3. In the sidebar, open the styles tab
4. Tweak the background color by opening the color picker. Drag the handle around the page, on top of the editor canvas
5. Make sure that the color picker's popover never closes unexpectedly, and that the handle follows the cursor around nicely while dragging.
6. While dragging, you may even try to move away from the browser window and release the pointer. The backdrop should disappear once re-entering the window with the cursor
7. While dragging, press the ESC key. The popover should close, and the backdrop should disappear too.

### Testing Instructions for Keyboard


## Screenshots or screencast 

### Before


https://github.com/WordPress/gutenberg/assets/1083581/cb8b5770-10b9-4ea9-b5f3-11e502e56f83

### After (note: the backdrop is purposefully tinted to better showcase the solution proposed)

https://github.com/WordPress/gutenberg/assets/1083581/80525f2d-4b81-4555-a2c1-9c184c7dd846
